src/EditorFeatures/CSharpTest/DocumentationComments/XmlTagCompletionTests.cs

```diff
@@ -251,9 +252,9 @@ class c { }";
             Verify(text, expected, '/');
         }
 
-        internal override ICommandHandler<TypeCharCommandArgs> CreateCommandHandler(ITextUndoHistoryRegistry undoHistory)
+        internal override IChainedCommandHandler<TypeCharCommandArgs> CreateCommandHandler(ITextUndoHistoryRegistry undoHistory)
```

Comment:
Is there a link to the docs on these types?

Reply:
I wish I could just publish the spec on GitHub :( Hopefully next features we will spec that way.

Essentially, the goal was to make simple handling simple and still allow advanced chaining scenarios. So we ended up with 2 ways to implement a command handler:
public interface ICommandHandler<T> : IDiagnosableCommandHandler where T : CommandArgs
{
    CommandState GetCommandState(T args);
    bool ExecuteCommand(T args, CommandExecutionContext executionContext);
}
and
public interface IChainedCommandHandler<T> : IDiagnosableCommandHandler where T : CommandArgs
{        
    CommandState GetCommandState(T args, Func<CommandState> nextCommandHandler);
    void ExecuteCommand(T args, Action nextCommandHandler, CommandExecutionContext executionContext);
}

IDiagnosableCommandHandler simply adds a DisplayName for diagnostics:
public interface IDiagnosableCommandHandler : ICommandHandler
{
        /// <summary>
        /// Gets display name of the command handler used to represent it to the user, for
        /// example when blaming it for delays or for commanding diagnostics.
        /// </summary>
        string DisplayName { get; }
}

IChainedCommandHandler works just as before, but ICommandHandler has no access to the next handler so the commanding service takes care of calling its next handler if needed (specifically when GetCommandState() returns CommandState.Undetermined or ExecuteCommand() returns false).

Reply:
good point